spec/test_collector/session_spec.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+RSpec.describe Buildkite::TestCollector::Session do
+  subject { described_class.new }
+  let(:data) do
+    {
+      "test-1": "test-1-data",
+      "test-2": "test-2-data",
+      "test-3": "test-3-data",
+      "test-4": "test-4-data",
+      "test-5": "test-5-data",
+    }
+  end
+
+  before do
+    Buildkite::TestCollector.uploader = Buildkite::TestCollector::Uploader
+    data.each { |k, v| Buildkite::TestCollector.uploader.traces[k] = v }
+    Buildkite::TestCollector.batch_size = 5
+  end
+
+  describe "#add_example_to_send_queue" do
+    it "adds the example and sends only when batch size is reached" do
+      expect(Buildkite::TestCollector::Uploader).to receive(:upload).once.with(data.values)
+
+      subject.add_example_to_send_queue(:"test-1")
+      subject.add_example_to_send_queue(:"test-2")
+      subject.add_example_to_send_queue(:"test-3")
+      subject.add_example_to_send_queue(:"test-4")
+      subject.add_example_to_send_queue(:"test-5")
+    end
+  end
+
+  describe "#send_remaining_data" do
+    it "sends through remaining examples even when batch size is not reach" do
+      expect(Buildkite::TestCollector::Uploader).to receive(:upload).once.with(data.reject { |k,v| k == :"test-5"}.values)
+
+      subject.add_example_to_send_queue(:"test-1")
+      subject.add_example_to_send_queue(:"test-2")
+      subject.add_example_to_send_queue(:"test-3")
+      subject.add_example_to_send_queue(:"test-4")
+
+      subject.send_remaining_data
+    end
+
+    it "limits upload batch size to UPLOAD_API_MAX_RESULTS" do
+      stub_const("Buildkite::TestCollector::Session::UPLOAD_API_MAX_RESULTS", 1)
+
+      expect(Buildkite::TestCollector::Uploader).to receive(:upload).twice
```

Comment:
Should we verify the data upload receives each time?

Reply:
Maybe? Not sure how easy this is to do tbh, I don't think you can chain a `with` call with two different sets of arguments after a `twice`? Plus we're already verifying the data in the previous test so 🤷‍♀️

Reply:
I think it's as easy as writing two separate expectations?

My thinking behind testing the arguments is to verify there isn't a bug with the incorrect executions being upload each time it is called.

I don't think the previous specs test for the case when the batch size is reached prematurely before `send_remaining_data` is called.